lib/client/data-manager-spec.js

```diff
@@ -0,0 +1,249 @@
+
+    sinon.assert.calledOnce(this.mockDataSetNotificationObservableStream.forEach);
+
+    assert.strictEqual(this.mockDataSetNotificationObservableStream, dataManager.stream);
+    assert.strictEqual(mock$fh, dataManager.$fh);
+    assert.strictEqual(mediator, dataManager.mediator);
+    assert.strictEqual(mockDataSetId, dataManager.datasetId);
+  });
+
+  it("should initialise without a mediator", function() {
+    var mock$fh = {};
+
+    var dataManager = new DataManager(mockDataSetId, mock$fh, this.mockDataSetNotificationObservableStream);
+
+    //It should not have attached a `forEach` filter to the observable stream if there is no mediator
+    sinon.assert.notCalled(this.mockDataSetNotificationObservableStream.forEach);
+
+    assert.strictEqual(this.mockDataSetNotificationObservableStream, dataManager.stream);
+    assert.strictEqual(mock$fh, dataManager.$fh);
+    assert.strictEqual(undefined, dataManager.mediator);
+    assert.strictEqual(mockDataSetId, dataManager.datasetId);
+  });
+
+  describe("CRUDL Operations", function() {
+
+    it("should list all items", function(done) {
+
+      //An example of a list API Response From the $fh.sync framework
+      var mockSyncDataSetListAPIResponse = {
+        "dataentryid": {
+          data: {
+            id: "SomeDataItemId"
+          },
+          hash: "somedatahashforthedataset"
+        }
+      };
+
+      var mock$fh = {
+        sync: {
+          doList: sinon.stub().callsArgWith(1, mockSyncDataSetListAPIResponse)
+        }
+      };
+
+      var dataManager = new DataManager(mockDataSetId, mock$fh, this.mockDataSetNotificationObservableStream, mediator);
+
+      dataManager.list().then(function(dataSetList) {
+
+        sinon.assert.calledWith(mock$fh.sync.doList, sinon.match(mockDataSetId), sinon.match.func, sinon.match.func);
+
+        assert.strictEqual(mockSyncDataSetListAPIResponse.dataentryid.data, dataSetList[0]);
+
+        done();
+      }).catch(done);
+    });
+
+    it("should create a new item", function(done) {
+
+      var mockRecordUid = "syncRecordUID";
+
+      var mockDataItem = {
+        name: "mynewdataitem",
+        description: "My New Data Item"
+      };
+
+      var mockSyncAPIResponse = {
+        uid: mockRecordUid
+      };
+
+      var expectedCreatedData = _.extend({_localuid: mockSyncAPIResponse.uid}, mockDataItem);
+
+      var mock$fh = {
+        sync: {
+          doCreate: sinon.stub().callsArgWith(2, mockSyncAPIResponse),
+          doUpdate: sinon.stub().callsArgWith(3, mockSyncAPIResponse),
+          doRead: sinon.stub().callsArgWith(2, _.extend({data: mockDataItem}, mockSyncAPIResponse))
+        }
+      };
+
+      var dataManager = new DataManager(mockDataSetId, mock$fh, this.mockDataSetNotificationObservableStream, mediator);
+
+      dataManager.create(mockDataItem).then(function(createResult) {
+        assert.ok(_.isEqual(expectedCreatedData, createResult), "Expected the created item with the local ID to be resolved");
```

Comment:
Does chai give nice diffs? Nice to know, It's been a while since I've used it. There's [`assert.equal`](https://nodejs.org/api/assert.html#assert_assert_equal_actual_expected_message) but it's probably not as nice.